### PR TITLE
table: Add simple content table

### PR DESCRIPTION
**Pull request by ChengYanJin**
**Component**: table 

**Description**:
Try to render the simple table content for the pod list table in Metalk8s UI.
image

Remain issue to solve: 
The scrollbar breaks the alignment of the header and body :(
still trying to figure out the solution. 

**Design**:
https://docs.google.com/document/d/1gZgZtcJtUZQlpN7H4zJhyeDK-c2w6y2UlaPGyGdqTpk/edit
PoC branch: `improvement/poc-new-table-component`

export the component in the next.js
`import { Table } from '@scality/core-ui/dist/next';`

**Breaking Changes**:

- [] Breaking Changes

---

Closes: #ISSUE_NUMBER

### CODE REVIEW

File: src/lib/components/tablev2/SimpleContent.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  TableBody,
+  TableHeader,
+  SortCaretWrapper,
+  SortIncentive,
+} from './Tablestyle';
+import { TableContext } from './Tablev2.component';
+
+export type SimpleContentProps = {
+  rowHeight: number,
+  outerTableHeight: number, //in pixel, the sum of height outside the table, pass it to virtualized table to calculate the height of the table
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: Can we compute this in the component ? Maybe we can wrap the table content with a div with display flex so that it fills parent height and use element.getBoundingClientRect() to retrieve the height.
This behavior could be shared with the multi select content so it could make sense to be extracted to a component that won't be exported (implementation detail of SimpleContent / MutliSelectContent components)

File: src/lib/components/tablev2/SimpleContent.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  TableRow,
+  TableBody,
+  TableHeader,
+  SortCaretWrapper,
+  SortIncentive,
+} from './Tablestyle';
+import { TableContext } from './Tablev2.component';
+
+export type SimpleContentProps = {
+  rowHeight: number,
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: Do we still need to give the control over the row height ? If so I think providing a set of options would be a better approach in order to unify the height across the UIs (a bit like the spacing component in a sense, provide a well-known pre defined scale). 
Additionally I think this could be an optional prop and default to a single line row height.

File: src/lib/components/tablev2/SimpleContent.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        <TableRow
+          {...row.getRowProps({
+            /* Note:
+            We need to pass the style property to the row component.
+            Otherwise when we scroll down, the next rows are flashing because they are re-rendered in loop. */
+            style: { ...style },
+          })}
+          row={row}
+          separationLineVariant={separationLineVariant}
+          className="tr"
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: Is this className required ? To me it could makes sense that the TabelRow component renders to an actual tr component

File: src/lib/components/tablev2/SimpleContent.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                ...cell.column.cellStyle,
+                // Vertically center the text in cells.
+                display: 'flex',
+                flexDirection: 'column',
+                justifyContent: 'center',
+              },
+            });
+
+            return (
+              <div {...cellProps} className="td">
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: I would suggest to switch to a real td component `<td {...cellProps}>`

File: src/lib/components/tablev2/SimpleContent.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+          })}
+        </TableRow>
+      );
+    },
+    [prepareRow, rows, separationLineVariant],
+  );
+
+  return (
+    <>
+      <div className="thead">
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: I would suggest to switch to a real thead component <thead>

File: src/lib/components/tablev2/SimpleContent.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        {headerGroups.map((headerGroup) => (
+          <HeadRow
+            {...headerGroup.getHeaderGroupProps()}
+            style={{
+              display: 'flex',
+            }}
+            separationLineVariant={separationLineVariant}
+            className="tr"
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: Same here HeadRow could render to a tr

File: src/lib/components/tablev2/SimpleContent.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  );
+
+  return (
+    <>
+      <div className="thead">
+        {headerGroups.map((headerGroup) => (
+          <HeadRow
+            {...headerGroup.getHeaderGroupProps()}
+            style={{
+              display: 'flex',
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: Can we move this style to styled-component definition of HeadRow ?

File: src/lib/components/tablev2/SimpleContent.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        {...getTableBodyProps()}
+        className="tbody"
+        outerTableHeight={outerTableHeight}
+      >
+        <AutoSizer>
+          {({ height, width }) => (
+            <List
+              height={height}
+              itemCount={rows.length} // how many items we are going to render
+              itemSize={rowHeight} // height of each row in pixel
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: I understand the List component from react-window takes this value in pixel but in our component we should compute a value taking in account the screen size to ease readability on monitoring screens and tablets.

File: src/lib/components/tablev2/TableUtil.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const STATUS_WARNING = 'warning';
+const STATUS_CRITICAL = 'critical';
+const STATUS_NONE = 'none';
+const STATUS_HEALTH = 'healthy';
+
+export const compareHealth = (status1, status2) => {
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/TableUtil.js]: Please add some typing here and unit test it

Comment (JBWatenbergScality) [reply src/lib/components/tablev2/TableUtil.js]: Passing an invalid status will result in `undefined - undefined` which will give NaN to the user. Throwing an error if the provided status are invalid would enhance the developer experience

File: src/lib/components/tablev2/Tablestyle.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  height: ${(props) => props.headerHeight};
+  &:hover {
+    ${SortIncentive} {
+      display: block;
+    }
+  }
+`;
+
+export const HeadRow = styled.div`
+  height: 32px;
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablestyle.js]: This should change depending on the screen height and could be considered as a minimum value eventually

Comment (JBWatenbergScality) [reply src/lib/components/tablev2/Tablestyle.js]: We probably want to derive this value form the font size of the headers and add some margin around instead of a fixed value

File: src/lib/components/tablev2/Tablev2.component.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+import SimpleContent from './SimpleContent';
+import { compareHealth } from './TableUtil.js';
+
+export type TableProps = {
+  columns: {
+    Header: string,
+    accessor: string,
+    sortType?:
+      | string
+      | ((rowA: any, rowB: any, columnId: string, desc: boolean) => number),
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: any can be replaced by generics deduced from the type of `data`

File: src/lib/components/tablev2/Tablev2.component.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    cellStyle?: { [key: string]: any },
+    Cell?: (cellProps: any) => any,
+  }[],
+  defaultSortingKey: string, //we don't display the default sort key in the URL, so we need to specify here
+  data: { [key: string]: any }[],
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: I would type it using `ROW[]` with `ROW` a generic type

File: src/lib/components/tablev2/Tablev2.component.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+import { compareHealth } from './TableUtil.js';
+
+export type TableProps = {
+  columns: {
+    Header: string,
+    accessor: string,
+    sortType?:
+      | string
+      | ((rowA: any, rowB: any, columnId: string, desc: boolean) => number),
+    cellStyle?: { [key: string]: any },
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: Use React.CSSProperties type

File: src/lib/components/tablev2/Tablev2.component.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    Cell?: (cellProps: any) => any,
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: It returns a React.Node

File: src/lib/components/tablev2/Tablev2.component.js [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+  }[],
+  defaultSortingKey: string, //we don't display the default sort key in the URL, so we need to specify here
+  data: { [key: string]: any }[],
+  rowIDKey: string, // define which column represent the unique id of the table
+  children: React.Node,
+  isMultiRowSelection?: boolean,
+  isCollapsible?: boolean,
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: I would remove those props for now as we decided to control multi row selection with a different child component and they are not yet implemented

File: src/lib/components/tablev2/SimpleContent.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+  rowHeight: number,
+  outerTableHeight: number, //in pixel, the sum of height outside the table, pass it to virtualized table to calculate the height of the table
+  separationLineVariant:
+    | 'backgroundLevel1'
+    | 'backgroundLevel2'
+    | 'backgroundLevel3'
+    | 'backgroundLevel4',
+};
+
+export default function SimpleContent({
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SimpleContent.js]: rename this to SingleSelectableContent to match the tech design

File: src/lib/components/tablev2/Tablev2.component.js [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  rows: rows,
+  prepareRow: prepareRow,
+  selectedRowIds: selectedRowIds,
+  selectedFlatRows: selectedFlatRows,
+  preGlobalFilteredRows: preGlobalFilteredRows,
+  setGlobalFilter: setGlobalFilter,
+  globalFilter: globalFilter,
+  setFilter: setFilter,
+};
+export const TableContext = React.createContext<TableContextValue>({});
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: A common practice is to initialize context with a null value in this way we can wrap their usage with a custom hook that check their value and throw an error if they are used outside of the context provider.

File: src/lib/components/tablev2/Tablev2.component.js [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+  children,
+  ...rest
+}: TableProps) {
+  const sortTypes = {
+    health: (row1, row2) => {
+      return compareHealth(row2.values.health, row1.values.health);
+    },
+  };
+
+  const getRowId = (row, relativeIndex) => {
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: nit: unify case of Id. In the props we have `rowIDKey` and here `getRowId`

Comment (JBWatenbergScality) [reply src/lib/components/tablev2/Tablev2.component.js]: I would opt to get this function from the props directly in order to give more control to the developer and address larger cases

File: src/lib/components/tablev2/Tablev2.component.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    return row[rowIDKey];
+  };
+
+  // Use the state and functions returned from useTable to build your UI
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: We can remove this comment

File: src/lib/components/tablev2/Tablev2.component.js [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    },
+    useBlockLayout,
+    useFilters,
+    useGlobalFilter,
+    useSortBy,
+    useExpanded,
+    useRowSelect,
+  );
+
+  // Render the UI for your table
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: We can remove this comment

File: src/lib/components/tablev2/Tablev2.component.js [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+        prepareRow,
+        selectedRowIds,
+        selectedFlatRows,
+        preGlobalFilteredRows,
+        setGlobalFilter,
+        globalFilter,
+        setFilter,
+      }}
+    >
+      {/* we need to use <div/> because of the virtualized table*/}
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/Tablev2.component.js]: is this due to a styling issue ? Can a css reset on the table element address this issue ? To me if possible it will be better to use a table component to keep the html semantic and benefit from browser rendering optimisations

**Review by JBWatenbergScality — CHANGES_REQUESTED** [review]

File: src/lib/components/tablev2/SimpleContent.js (hunk 2 above)
Comment (ChengYanJin) [reply src/lib/components/tablev2/SimpleContent.js]: After checking the design from @Cuervino , I see the following possibilities.
```js
export const tableRowHeight = {
  h32: '2.286rem',
  h40: '2.858rem',
  h48: '3.428rem',
  h64: '4.572rem',
};
```
but I am not sure if we want to have the default row height => to be discussed.

Comment (Cuervino) [reply src/lib/components/tablev2/SimpleContent.js]: IMO, offering those 4 predefined row height is rather good.
From a visual design point of view, these multiple of 8-px give good balance between components.

File: src/lib/components/tablev2/SimpleContent.js (hunk 1 above)
Comment (ChengYanJin) [reply src/lib/components/tablev2/SimpleContent.js]: It's a great idea. But unfortunately, without explicitly define the height of the table body, the height will be 0, hence no content can be displayed.

I don't see any other way than expose a property to ask the developer to calculate the sum height of outerTableBody.

File: src/lib/components/tablev2/SimpleContent.js (hunk 8 above)
Comment (ChengYanJin) [reply src/lib/components/tablev2/SimpleContent.js]: so we maybe should define the `rowHeight` in `vh` and convert it to pixel?

File: src/lib/components/tablev2/Tablestyle.js (hunk 10 above)
Comment (ChengYanJin) [reply src/lib/components/tablev2/Tablestyle.js]: for the moment, I will just change it to `height: 2.286rem;` and vertical aligned

File: src/lib/components/tablev2/SingleSelectableContent.js [hunk 22]
```diff
@@ -10,32 +10,50 @@ import {
...
+  if (['h32', 'h40', 'h48', 'h64'].indexOf(rowHeight) === -1) {
+    console.error('Invalid table row height defined.');
+  }
+  if (
+    [
+      'backgroundLevel1',
+      'backgroundLevel2',
+      'backgroundLevel3',
+      'backgroundLevel4',
+    ].indexOf(separationLineVariant) === -1
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SingleSelectableContent.js]: can be replaced by `includes`

File: src/lib/components/tablev2/SingleSelectableContent.js [hunk 23]
```diff
@@ -10,32 +10,50 @@ import {
...
+  ) {
+    console.error('Invalid table seperation line color defined.');
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/SingleSelectableContent.js]: I would reword the error message to `Invalid separationLineVariant props, expected backgroundLevel1, backgroundLevel2, backgroundLevel3 or backgroundLevel4 but received ${separationLineVariant}`

File: src/lib/components/tablev2/TableUtil.js [hunk 24]
```diff
@@ -1,14 +1,39 @@
...
 const STATUS_HEALTH = 'healthy';
 
-export const compareHealth = (status1, status2) => {
+// some common customized sortTypes
+export function compareHealth(
+  status1: STATUS_WARNING | STATUS_CRITICAL | STATUS_NONE | STATUS_HEALTH,
+  status2: STATUS_WARNING | STATUS_CRITICAL | STATUS_NONE | STATUS_HEALTH,
+): number {
+  if (
+    [STATUS_WARNING, STATUS_CRITICAL, STATUS_NONE, STATUS_HEALTH].indexOf(
```
Comment (JBWatenbergScality) [inline src/lib/components/tablev2/TableUtil.js]: Can be replaced by `includes`

File: src/lib/components/tablev2/SingleSelectableContent.js (hunk 23 above)
Comment (ChengYanJin) [reply src/lib/components/tablev2/SingleSelectableContent.js]: Yes. indeed. it's more developer friendly ;)

**Review by JBWatenbergScality — APPROVED** [review]
